auth/src/main/java/com/firebase/ui/auth/viewmodel/idp/EmailProviderResponseHandler.java

```diff
@@ -0,0 +1,103 @@
+
+import android.app.Application;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.firebase.ui.auth.IdpResponse;
+import com.firebase.ui.auth.data.model.IntentRequiredException;
+import com.firebase.ui.auth.data.model.Resource;
+import com.firebase.ui.auth.data.model.User;
+import com.firebase.ui.auth.data.remote.ProfileMerger;
+import com.firebase.ui.auth.ui.email.WelcomeBackPasswordPrompt;
+import com.firebase.ui.auth.ui.idp.WelcomeBackIdpPrompt;
+import com.firebase.ui.auth.util.data.ProviderUtils;
+import com.firebase.ui.auth.util.data.TaskFailureLogger;
+import com.firebase.ui.auth.viewmodel.AuthViewModelBase;
+import com.firebase.ui.auth.viewmodel.RequestCodes;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.EmailAuthProvider;
+import com.google.firebase.auth.FirebaseAuthUserCollisionException;
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class EmailProviderResponseHandler extends AuthViewModelBase<IdpResponse> {
+    private static final String TAG = "EmailProviderResponseHa";
+
+    public EmailProviderResponseHandler(Application application) {
+        super(application);
+    }
+
+    public void startSignIn(@NonNull final IdpResponse response, @NonNull String password) {
+        if (!response.isSuccessful()) {
+            setResult(Resource.<IdpResponse>forFailure(response.getError()));
+            return;
+        }
+        if (!response.getProviderType().equals(EmailAuthProvider.PROVIDER_ID)) {
+            throw new IllegalStateException(
+                    "This handler can only be used with the email provider");
+        }
+        setResult(Resource.<IdpResponse>forLoading());
+
+        final String email = response.getEmail();
+        getAuth().createUserWithEmailAndPassword(email, password)
+                .continueWithTask(new ProfileMerger(response))
+                .addOnFailureListener(new TaskFailureLogger(TAG, "Error creating user"))
+                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
+                    @Override
+                    public void onSuccess(AuthResult result) {
+                        setResult(Resource.forSuccess(response));
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        if (e instanceof FirebaseAuthUserCollisionException) {
+                            // Collision with existing user email, it should be very hard for
+                            // the user to even get to this error due to CheckEmailFragment.
+                            ProviderUtils.fetchTopProvider(getAuth(), email)
+                                    .addOnSuccessListener(new StartWelcomeBackFlow(email));
```

Comment:
What about when this call fails too?  Feels like we could get a case where the `Resource` never gets to a final state.

Reply:
Whoops, yeah.